File: Sean的leetcode/FB/Binary Tree Vertical Order Traversal.py
###
# BFS: ensure that upper level is printed before lower level
# Time Complexity: O(nlogn) (sorted dict)
# Space Complexity: O(n)
###
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def verticalOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        q = [(root, 0)]
        table = collections.defaultdict(list)
        for node, i in q:
            table[i].append(node.val)
            if node.left:
                q.append((node.left, i-1))
            if node.right:
                q.append((node.right, i+1))
        return [table[i] for i in sorted(table)]

File: Sean的leetcode/Leetcode/Binary Tree Vertical Order Traversal.cpp
//BFS
//Time Complexity: O(nlogn) (sorted map)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> verticalOrder(TreeNode* root) {
        vector<vector<int>> res;
        if (!root) return res;
        queue<pair<TreeNode*, int>> q;
        q.push({root, 0});
        map<int, vector<int>> table;
        while (!q.empty()) {
            TreeNode* nd = q.front().first;
            int col = q.front().second;
            q.pop();
            table[col].push_back(nd->val);
            if (nd->left) {
                q.push({nd->left, col - 1});
            }
            if (nd->right) {
                q.push({nd->right, col + 1});
            }
        }
        for (auto e: table) {
            res.push_back(e.second);
        }
        return res;
    }
};